scenario/context.py

```diff
@@ -34,7 +34,25 @@
 
 logger = scenario_logger.getChild("runtime")
 
-ActionOutput = namedtuple("ActionOutput", ("state", "logs", "results", "failure"))
+
+@dataclasses.dataclass
+class ActionOutput:
+    """Wraps the results of running an action event with `run_action`."""
+
+    state: "State"
+    """The charm state after the action has been handled.
+    In most cases, actions are not expected to be affecting it."""
+    logs: List[str]
+    """Any logs associated with the action output, set by the charm."""
+    results: Dict[str, str]
```

Comment:
Wouldn't this be `Dict[str, Any]` to match `ActionEvent.set_results`? ops flattens the results down to a string, but then Juju unflattens them out again when displaying them.

Reply:
mmm yes, you're right. will fix